README.md
# ProbeAPI Tools
## About
`ProbeAPI Tools` is a set of cross platform console command line utilities which are using [public API](http://www.probeapi.com/) to make distributed pinging, tracerouting and etc.
All these tools are working almost the same as Windows and Linux built-in tools, but all network operations are going from different hosts in the internet (we call them probes).
"d" prefix from the name of all tools comes from the word **d**istributed.

## Particular Tools Documentation

### Windows
* [d**ping**](doc/dping.windows.md)
* [d**tracert**](doc/dtraceroute.windows.md)

### Linux
* [d**ping**](doc/dping.linux.md)
* [d**traceroute**](doc/dtraceroute.linux.md)

## ChangeLog
[History of changes of ProbeAPI Tools](ChangeLog.md)

## Compilation

Two main ways to build tools are supported:
* Microsoft Visual Studio 2013 solution
* nmake config (CMakeLists.txt)

Tested environments and compilers:
* Microsoft Visual Studio 2013 (2012 should be OK too, but not 2010)
* MinGW for Windows (with gcc 4.8.1)
* gcc 4.9.2 (tested in Ubuntu 15.04)

### Compilation in Linux
In Linux you will need to install some packages prior to building projects. Please run the following commands to install required packages:
```
sudo apt-get install git cmake libcurl4-openssl-dev make gcc
```
Installing optional packages (needed in case of building documentation, building and installing RPM):
```
sudo apt-get install librpmbuild3 alien bash gawk
```

Checkout and build ProbeAPI Tools:
```
cd ~
git clone --recursive https://github.com/optimal-software/probe-api.git
cd ~/probe-api/build
cmake ..
make
```

Now you can run tools:
```
cd ~/probe-api/bin
./dping --version
./dtraceroute --version
```

To install tools after building:
```
cd ~/probe-api/build
sudo make install
```

You can also build RPM:
```
cd ~/probe-api/build
make package
```

### Compilation in Windows

Please checkout repository recursively:
```
git clone --recursive https://github.com/optimal-software/probe-api.git
```

#### MinGW
1. Please install cmake for Windows.
2. Execute following commands:
```
cd ~
git clone --recursive https://github.com/optimal-software/probe-api.git
cd ~/probe-api/build
```
And later:
```
cmake -G "MinGW Makefiles" ..
make
```
or
```
cmake ..
make
```
To install after building:
```
sudo make install
```

#### Microsoft Visual Studio

Microsoft Visual Studio solution file is provided in the root of Git repository.

## Installation

No special installation is needed for any of the tools. Currently precompiled binary EXE files are provided and they do not need any installation. Executables are fully portable and they have no external dependencies.
Pre-built 64 bit RPM is also provided for Linux operation systems. It may be installed by command:
```
sudo alien -i probe-api-tools-******.x86_64.rpm
```
or
```
sudo rpm -i probe-api-tools-******.x86_64.rpm
```
(please replace "******" by corresponding product version from RPM file name)

## Credits

```
Copyright (c) 2015 Optimal Software s.r.o. All rights reserved.

Sources repository:  https://github.com/optimal-software/probe-api
Latest license:      https://github.com/optimal-software/probe-api/blob/master/LICENSE

Build mode: Release

Authors: Sergey Kolomenkin

===============================================================================

The MIT License (MIT)

Copyright (c) 2015 ProbeAPI Tools

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

===============================================================================

Used third-party libraries:
cURL        : libcurl/7.42.1 WinSSL zlib/1.2.8; Protocols: http https
    Website : http://curl.haxx.se/
    Repo    : https://github.com/bagder/curl.git
    License : MIT; http://curl.haxx.se/docs/copyright.html
  zlib
    Website : http://zlib.net/
    Repo    : https://github.com/madler/zlib.git
    License : MIT; http://zlib.net/zlib_license.html
cURLpp      : 0.7.3
    Website : http://rrette.com/curlpp.html
    Repo    : https://github.com/jpbarrette/curlpp.git
    License : MIT; http://www.curlpp.org/#license
JsonCpp     : 0.10.2
    Website : https://github.com/open-source-parsers/jsoncpp
    Repo    : https://github.com/open-source-parsers/jsoncpp.git
    License : Public Domain, MIT; https://github.com/open-source-parsers/jsoncpp/blob/master/LICENSE
```


ChangeLog.md
# ProbeAPI Tools History (ChangeLog)

## Next version


## Version 1.4.34 (12 September 2016)

* moved from Mashape to Kong


## Version 1.4.33 (19 August 2015)

* minor changes in error messages
* Changed default values in dping:
  * single ping timeout: 3000ms -> 1000ms
  * wait time for server response: 2000ms -> 4000ms
* Changed default values in dtracert:
  * wait time for server response: 2000ms -> 4000ms


## Version 1.4.32 (27 July 2015)

* added command line options for dtracert:
  * -hs hop_number  First hop number to start from.


## Version 1.4.30 (29 June 2015)

* updated code to use and parse latest web API changes: more command line options, more control and etc.
* added command line options for both dping and dtracert:
  * -l size         - Send buffer size.
  * -f              - Set Don't Fragment flag in packet (IPv4-only).
  * -4              - Force using IPv4.
  * -6              - Force using IPv6.
  * --api-key key   - Set web API key.
  * --api-url url   - Set web API URL.
* added command line options for dping: -t, -a, -np, -i
  * -t              - Ping the specified host until stopped. To see statistics and continue - type Control-Break; To stop - type Control-C.
  * -a              - Resolve addresses to hostnames.
  * -np count       - Number of pings per probe to send (default: 1).
  * -i TTL          - Time To Live.
* added command line options for dtracert:
  * -d              - Do not resolve addresses to hostnames.
  * -hf max_hops    - Maximum number of failed hops in a row to stop (default: 3).
* changed default dping TTL from 128 to 55 (like in Windows 8)
* Changed default values in dping:
  * single ping timeout: 5000 ms -> 3000 ms
  * pings per probe: 3 -> 1
  * wait between pings: 1000ms -> 100 ms
  * TTL: 128 -> 55
  * total time for API call: 5000ms (unchanged)
  * probe count: 10 (unchanged)
* Changed default values in dtracert:
  * single ping timeout: 3000 ms -> 1000 ms
  * wait between pings: 1000ms -> 100 ms
  * total time for API call: 44 sec -> 47 sec
  * probe count: 40 (default 10 * factor 4) -> 10
* fixed bug with delays during dtraceroute print. Printing is much faster now
* don't exit if Control+Break is catched during dping. Print statistics and continue (like Windows ping does)
* allow target host to be specified at any position of command line (not necessary to have it at the end of command line any more)
* added --probes command line switch as an alias for -n (both dping and dtracert)
* other fixes


## Version 1.3.28 (6 June 2015)

* added --no-delays command line option
* use zlib for cURL in Windows
* Minor bugs fixed
* added cmake scripts for crossplatform building
* supported compilation environments:
  * native MSVC
  * cmake: auto-generated MSVC
  * cmake: MinGW for Windows + makefiles + gcc
  * cmake: Unix makefiles + gcc in Linux
* use submodules in git to get dependent sources
* redistribute pre-built RPM packet for linux


## Version 1.1.22 (22 May 2015)

* fixed a bug in dping which caused skipping of final statistics print
* documentation is generated automatically now
* other code improvements

**WARNING!** dtraceroute's --asn mode is broken in this release!

 
## Version 1.1.20 (21 May 2015)

* dtraceroute project is implemented
* program termination is now more correct
* source code was deduplicated


## Version 1.0.16 (18 May 2015)

* Added license, copyrights, documentation.
* Disabled default mode which caused pinging from country with most of endpoints online.
* Smaller curl runtime (excluded unused protocols).


## Version 1.0.13 (13 May 2015)

All dping utility features from wiki are implemented:

* --list-country - List available countries.
* --list-asn {code} - List available ASNs for specified country code (ISO 3166-1 alpha-2).
* --country {code} - Ping from specified 2 letter country code (ISO 3166-1 alpha-2).
* --asn {id} - Ping from specified ASN.

Usage examples:
* dping --asn AS12357 8.8.8.8
* dping --country US 8.8.8.8


## Version 0.5.7 (11 May 2015)

dping utility got its first few features implemented.

It currently allows two main command line modes:

* --list-country - List available countries.
* --country {code} - Ping from specified 2 letter country code (ISO 3166-1 alpha-2).


src/dping/dping.cpp
//------------------------------------------------------
#include "stdafx.h"
#include "dping.h"

#include "common/Common.h"
#include "common/CommonOperations.h"
#include "common/ParseReply.h"

using namespace std;

#ifdef DEST_OS_WINDOWS
#define PRINT_AS_WINDOWS
#endif

//------------------------------------------------------

class JobType;
shared_ptr<JobType> g_ptrJob = nullptr;
shared_ptr<ApplicationStats> g_ptrStats = nullptr;

//------------------------------------------------------
// Windows sample: (Win 8.1)

// c:\bin>ping 8.8.8.8
// 
// Pinging 8.8.8.8 with 32 bytes of data:
// Reply from 8.8.8.8: bytes=32 time=14ms TTL=55
// Reply from 8.8.8.8: bytes=32 time=13ms TTL=55
// Reply from 8.8.8.8: bytes=32 time=13ms TTL=55
// Reply from 8.8.8.8: bytes=32 time=13ms TTL=55
// 
// Ping statistics for 8.8.8.8:
// Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
// Approximate round trip times in milli-seconds:
// Minimum = 13ms, Maximum = 14ms, Average = 13ms
// 
// c:\bin>

//------------------------------------------------------
// Linux sample: (ubuntu)

// sergey@ubuntu:~$ ping 8.8.8.8
// PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
// 64 bytes from 8.8.8.8: icmp_seq=1 ttl=128 time=13.7 ms
// 64 bytes from 8.8.8.8: icmp_seq=2 ttl=128 time=14.0 ms
// 64 bytes from 8.8.8.8: icmp_seq=3 ttl=128 time=15.7 ms
// 64 bytes from 8.8.8.8: icmp_seq=4 ttl=128 time=15.4 ms
// 64 bytes from 8.8.8.8: icmp_seq=5 ttl=128 time=15.4 ms
// 64 bytes from 8.8.8.8: icmp_seq=6 ttl=128 time=13.5 ms
// 64 bytes from 8.8.8.8: icmp_seq=7 ttl=128 time=14.4 ms
// 64 bytes from 8.8.8.8: icmp_seq=8 ttl=128 time=14.9 ms
// 64 bytes from 8.8.8.8: icmp_seq=9 ttl=128 time=14.3 ms
// 64 bytes from 8.8.8.8: icmp_seq=10 ttl=128 time=13.8 ms
// 64 bytes from 8.8.8.8: icmp_seq=11 ttl=128 time=18.1 ms
// 64 bytes from 8.8.8.8: icmp_seq=12 ttl=128 time=14.6 ms
// 64 bytes from 8.8.8.8: icmp_seq=13 ttl=128 time=16.4 ms
// ^C
// --- 8.8.8.8 ping statistics ---
// 13 packets transmitted, 13 received, 0% packet loss, time 12025ms
// rtt min/avg/max/mdev = 13.521/14.979/18.114/1.232 ms
// sergey@ubuntu:~$

// sergey@ubuntu:~$ ping 1.1.1.2
// PING 1.1.1.2 (1.1.1.2) 56(84) bytes of data.
// ^C
// --- 1.1.1.2 ping statistics ---
// 33 packets transmitted, 0 received, 100% packet loss, time 32059ms
// 
// sergey@ubuntu:~$

//------------------------------------------------------

class JobType
{
public:
	JobType(const ApplicationOptions& options_) : options(options_)
	{
		switch (options.mode)
		{
		case ApplicationOptions::MODE_DO_BY_COUNTRY:
			sMethod = "StartPingTestByCountry";
			sSearchArgName = "countrycode";
			sResultRoot = "StartPingTestByCountryResult";
			sSearchDetails = "country code {ARG}";
			break;
		case ApplicationOptions::MODE_DO_BY_ASN:
			sMethod = "StartPingTestByAsn";
			sSearchArgName = "asnid";
			sResultRoot = "StartPingTestByASNResult";
			sSearchDetails = "{ARG}";
			break;
		default:
			sMethod = "does_not_exist";
			sSearchArgName = "does_not_exist";
			sResultRoot = "does_not_exist";
			sSearchDetails = "{ARG}";
			break;
		}
	}
	~JobType()
	{
	}

	string GetUrl(const ApplicationStats& stats, const string& sSearchArgument, const string& sTarget) const
	{
		// Documentation:
		// https://cloudperf.speedchecker.xyz/application/docs-probeapi/#ping-test-by-country
		// https://cloudperf.speedchecker.xyz/application/docs-probeapi/#ping-test-by-asn
		const int64_t nRestProbes = options.nProbesLimit - stats.nSent;
		const auto nRequestedProbeCount = (std::min)(options.nMaxProbesToRequest, (std::max)(options.nMinProbesToRequest, nRestProbes));

		const string sUrl = OSSFMT(sMethod
			<< "?" << sSearchArgName << "=" << sSearchArgument
			<< "&destination=" << sTarget					// An IP address or hostname that will be pinged
			<< "&probeslimit=" << nRequestedProbeCount		// Number of probes to use
			<< "&count=" << options.nPingsPerProbe			// Number of pings to run (default 3). (per each probe)
			<< "&ttl=" << options.nTTL						// Max number of hops for ping
			<< "&sleep=" << options.nWaitBetweenPingsMs		// Sleep between pings in milliseconds (default 1000ms).
			<< "&timeout=" << options.nTimeoutTotalMs		// Maximum time available to probes for testing in milliseconds (default 6000). The whole test is most likely to last longer then this value.
			<< "&commandTimeout=" << options.nTimeoutPingMs // Ping timeout in milliseconds that one ping can take (default 5000ms).
			<< "&bufferSize=" << options.nPacketSize		// buffer size filled with 'A' char to send, default=32, max=65500
			<< "&fragment=" << !options.bDontFragment		// opposite of "dontFragment" flag, default=1 => dontFragment=0
			<< "&resolve=" << options.bResolveIp2Name		// if IP was given, try to resolve it, default=0
			<< "&ipv4only=" << options.bUseIpv4Only			// force using IPv4 (if no IPv4 IP address is returned, return error), default=0
			<< "&ipv6only=" << options.bUseIpv6Only			// force using IPv6 (if no IPv6 IP address is returned, return error), default=0
			);

		return sUrl;
	}

	string GetResultRoot() const
	{
		return sResultRoot;
	}

	string FormatSearchDetails(const string& sSearchArgument) const
	{
		return findandreplaceConst(sSearchDetails, "{ARG}", sSearchArgument);
	}

	string CalculateSearchArgument(ProbeApiRequester& requester) const
	{
		string sSearchArgument = options.sModeArgument;
		if (ApplicationOptions::MODE_DO_BY_COUNTRY == options.mode && DEFAULT_COUNTRY_META == sSearchArgument)
		{
			const CommonOptions options2 = options.GetCommonOptions();
			sSearchArgument = GetDefaultSourceCountry(requester, options2);
		}
		return sSearchArgument;
	}

	void PrintHeaderBeforeSearchArg() const
	{
#ifdef PRINT_AS_WINDOWS
		// 
		// Pinging 8.8.8.8 with 32 bytes of data:
		// Pinging google-public-dns-a.google.com [8.8.8.8] with 32 bytes of data:
		cout << endl;
		cout << "Pinging " << options.sTarget << " with " << options.nPacketSize << " bytes of data";
#else
		// PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
		// PING ya.ru (213.180.193.3) 56(84) bytes of data.
		const int nHeadersSize = 20 + 8; // IP + ICMP headers size
		cout << "PING " << options.sTarget << " (" << options.sTarget << ") " << options.nPacketSize << "(" << (options.nPacketSize + nHeadersSize) << ") bytes of data";
#endif
	}

	void PrintHeaderAfterSearchArg(const string& sSearchArgument) const
	{
#ifdef PRINT_AS_WINDOWS
		cout << " from " << FormatSearchDetails(sSearchArgument) << ":" << endl;
#else
		cout << " from " << FormatSearchDetails(sSearchArgument) << "." << endl;
#endif
	}

	void PrintJobResult(const ProbeAPI::ProbeInfo& info, const ProbeAPI::PingResult& pingResult) const
	{
#ifdef PRINT_AS_WINDOWS
		// Reply from 8.8.8.8: bytes=32 time=13ms TTL=55
		if (pingResult.bTimeout)
		{
			cout << "Request timed out.";
		}
		else
		{
			const auto& remote = info.ping;
			const string sTargetInfo = options.sTarget == remote.sTargetHost ? remote.sTargetIp : remote.sTargetHost + " [" + remote.sTargetIp + "]";

			cout << "Reply from " << sTargetInfo << ": bytes=" << options.nPacketSize << " time=" << pingResult.nTimeMs << "ms TTL=" << options.nTTL;
		}

		if (options.bVerbose)
		{
			cout << " for " << info.GetProbeInfo(options.mode == ApplicationOptions::MODE_DO_BY_ASN);
		}

		cout << endl;
#else
		// 64 bytes from 8.8.8.8: icmp_seq=1 ttl=128 time=13.7 ms
		if (pingResult.bTimeout)
		{
		}
		else
		{
			const auto& remote = info.ping;
			const string sTargetInfo = options.sTarget == remote.sTargetHost ? remote.sTargetIp : remote.sTargetHost + " (" + remote.sTargetIp + ")";

			cout << options.nPacketSize << " bytes from " << sTargetInfo << ": icmp_seq=1 ttl=" << options.nTTL << " time=" << pingResult.nTimeMs << ".0 ms";
			if (options.bVerbose)
			{
				cout << " for " << info.GetProbeInfo(options.mode == ApplicationOptions::MODE_DO_BY_ASN);
			}
			cout << endl;
		}
#endif
	}

	void PrintFooter(const ApplicationStats& stats) const
	{
#ifdef PRINT_AS_WINDOWS
		// 
		// Ping statistics for 8.8.8.8:
		// Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
		// Approximate round trip times in milli-seconds:
		// Minimum = 13ms, Maximum = 14ms, Average = 13ms
		cout << endl;
		cout << "Ping statistics for " << options.sTarget << endl;
		cout << "    Packets : Sent = " << stats.nSent << ", Received = " << stats.nReceived << ", Lost = " << (stats.nSent - stats.nReceived)
			<< " (" << ((stats.nSent - stats.nReceived) * 100 / (stats.nSent ? stats.nSent : 1)) << "% loss)," << endl;

		if (stats.nReceived > 0)
		{
			cout << "Approximate round trip times in milli-seconds:" << endl;
			cout << "    Minimum = " << stats.pings.GetMin() << "ms, Maximum = "
				<< stats.pings.GetMax() << "ms, Average = " << (int)stats.pings.GetAverage() << "ms" << endl;
		}
#else
		// ^C
		// --- 8.8.8.8 ping statistics ---
		// 13 packets transmitted, 13 received, 0% packet loss, time 12025ms
		// rtt min/avg/max/mdev = 13.521/14.979/18.114/1.232 ms
		cout << "--- " << options.sTarget << " ping statistics ---" << endl;
		cout << stats.nSent << " packets transmitted, " << stats.nReceived << " received, "
			<< ((stats.nSent - stats.nReceived) * 100 / (stats.nSent ? stats.nSent : 1)) << "% packet loss, time "
			<< stats.GetTimeElapsedMs() << "ms" << endl;

		if (stats.nReceived > 0)
		{
			cout << "rtt min/avg/max/mdev = " << fixed << setprecision(3)
				<< (double)stats.pings.GetMin() << "/" << (double)stats.pings.GetAverage()
				<< "/" << (double)stats.pings.GetMax() << "/" << (double)stats.pings.GetMedianAbsoluteDeviation()
				<< resetiosflags(cout.flags()) << resetiosflags(ios_base::floatfield) << " ms" << endl;
		}
#endif
	}

protected:
	const ApplicationOptions&	options;
	string						sMethod;
	string						sSearchArgName;
	string						sSearchDetails;
	string						sResultRoot;
};

//------------------------------------------------------

void PrintFinalStats()
{
	shared_ptr<ApplicationStats> ptrStats = g_ptrStats;
	shared_ptr<JobType> ptrJob = g_ptrJob;

	if (ptrJob && ptrStats)
	{
		ptrJob->PrintFooter(*ptrStats);
	}
}

//------------------------------------------------------

void PrintPackOfResults(const JobType& job, const ApplicationOptions& options, const vector<ProbeAPI::ProbeInfo>& items, ApplicationStats& stats)
{
	//bool bFirstIteration = (0 == stats.nSent);
	bool bFirstIteration = true;
	for (const auto& info : items)
	{
		for (const auto& pingResult : info.ping.vectResults)
		{
			if (g_bTerminateProgram)
				throw PException("PrintPackOfResults: Terminate Program");

			++stats.nSent;

			if (!pingResult.bTimeout)
			{
				++stats.nReceived;
				stats.pings.AddItem(pingResult.nTimeMs);
			}

			if (!options.bNoDelays)
			{
				DoSleep(pingResult, bFirstIteration);
			}
			job.PrintJobResult(info, pingResult);

			if (stats.nSent == options.nProbesLimit)
				return;
		}
	}
}

//------------------------------------------------------

int MakePackOfJobs(const JobType& job, const string& sSearchArgument,
	const ApplicationOptions& options, ProbeApiRequester& requester, ApplicationStats& stats)
{
	const string sUrl = job.GetUrl(stats, sSearchArgument, options.sTarget);

	ProbeApiRequester::Request request(sUrl, options.GetCommonOptions());
	request.nHttpTimeoutSec += options.nTimeoutTotalMs / 1000;

	const ProbeApiRequester::Reply reply = requester.DoRequest(request, options.bDebug);
	if (!reply.bSucceeded)
	{
		if (reply.nHttpCode == 203)
			throw PException("Daily API limit reached - more information - http://www.probeapi.com/ip_api_limit_reached.html", eRetCode::ApiFailure);
		else
			throw PException(reply.sErrorDescription, eRetCode::ApiFailure);
	}

	vector<ProbeAPI::ProbeInfo> items;

	try
	{
		items = ProbeAPI::ParsePingResults(reply.sBody, job.GetResultRoot());
	}
	catch (PException& e)
	{
		throw PException("MakePackOfJobs: ProbeAPI::ParsePingResults: " + e.str(), eRetCode::ApiParsingFail);
	}

	PrintPackOfResults(job, options, items, stats);

	return eRetCode::OK;
}

//------------------------------------------------------

int DoJob(const ApplicationOptions& options)
{
	int res = eRetCode::OK;

	shared_ptr<ApplicationStats> ptrStats = make_shared<ApplicationStats>(options.sTarget);
	ApplicationStats& stats = *ptrStats;
	g_ptrStats = ptrStats;

	shared_ptr<JobType> ptrJob = make_shared<JobType>(options);
	const JobType& job = *ptrJob;
	g_ptrJob = ptrJob;

	job.PrintHeaderBeforeSearchArg();

	ProbeApiRequester requester;
	const string sSearchArgument = job.CalculateSearchArgument(requester);

	job.PrintHeaderAfterSearchArg(sSearchArgument);

	try
	{
		while (stats.nSent < options.nProbesLimit)
		{
			const auto nPreviousSend = stats.nSent;

			const int nRes = MakePackOfJobs(job, sSearchArgument, options, requester, stats);
			if (eRetCode::OK != nRes)
			{
				res = nRes;
				break;
			}
			if (nPreviousSend == stats.nSent)
			{
				// don't try again if no results are returned!
				break;
			}

			if (g_bTerminateProgram)
				throw PException("DoJob: Terminate Program");
		}
	}
	catch (...)
	{
		if (stats.nSent > 0)
			job.PrintFooter(stats);
		g_ptrStats.reset();
		g_ptrJob.reset();
		throw;
	}

	job.PrintFooter(stats);
	g_ptrStats.reset();
	g_ptrJob.reset();

	return res;
}

//------------------------------------------------------

int Application(const ApplicationOptions& options)
{
	const CommonOptions options2 = options.GetCommonOptions();

	switch (options.mode)
	{
	case ApplicationOptions::MODE_DO_BY_COUNTRY:
		return DoJob(options);
	case ApplicationOptions::MODE_DO_BY_ASN:
		return DoJob(options);
	case ApplicationOptions::MODE_GET_COUNTRIES:
		return ListCountries(options2);
	case ApplicationOptions::MODE_GET_ASNS:
		return ListAsns(options2);
	default:
		throw PException(eRetCode::NotSupported) << "Unknown program mode " << options.mode;
	}
}

//------------------------------------------------------
